Servers/FedBeServer.py

```diff
@@ -0,0 +1,121 @@
+        if self.verbose: print('FedBe: Reading in clients.')
+        mu_n = 0
+        s_n = 0
+        M_n = 0
+        S=[]
+        for n,client in enumerate(clients):
+            w = client.get_weights()
+            w_flat = torch.cat([w[k].flatten() for k in w])
+            mu_n2 = (n*mu_n+w_flat)/(n+1)
+            M_n= M_n+(w_flat-mu_n)*(w_flat-mu_n2)
+            if(n!=0):
+                s_n= torch.sqrt(M_n/n)
+            mu_n = mu_n2
+            S.append(w)
+
+        # Add small variance to avoid zero variance
+        s_n = s_n+0.0000001
+        self.distribution = Normal(mu_n,s_n)
+
+        if self.verbose: print('FedBe: Creating ensamble of weights.')
+        # Add mu
+        mu_r = torch.split(mu_n,self.tens_lengths)
+        mu_r = {k:torch.reshape(mu_r[i],self.model_shapes[i]) for i,k in enumerate(w)}
+        S.append(mu_r)
+        # Sample new weights
+        for i in range(self.M):
+            w_s = self.distribution.sample()
+            w_r = torch.split(w_s,self.tens_lengths)
+            w_r = {k:torch.reshape(w_r[i],self.model_shapes[i]) for i,k in enumerate(w)}
+            S.append(w_r)
+
+        ################ Evaluate on local data #######################
+        if self.verbose: print('FedBe: Evaluating ensambles on local data.')
+        loss = NLLLoss(reduction  ='sum')
+        self.model.eval()
+        p=np.array([])
+        X = []
+        for w in S:
+            self.model.set_weights(w)
+            res = []
+            for x,y in self.loc_data:
+                y_pred = self.model.forward(x)[0]
+                # Calculate likelihood per sample
+                for x,y_p,y_t in zip(x,y_pred,y):
+                    nll =loss(y_p,y_t).detach()
+                    l = torch.exp(-nll)
+                    res.append(l)
+                    # Add samples to map loss landscape only during first iteration
+                    if(len(X)!=len(p)):
+                        X.append(x)
+            res = np.array(res)
+            if(p.size == 0):
+                p = np.zeros(len(res))
+            p+=res
+
+        p = p/len(S)
+        T = [(x_j,p_j) for x_j,p_j in zip(X,p)]
+
+        ############## SWA ############################
+        if self.verbose: print('FedBE: Running SWA')
+        # Batch data
+        random.seed(self.seed)
+        random.shuffle(T)
+        T_batched = [T[i:((i+1)*self.swa_batch_size )] for i in range(len(T)//self.swa_batch_size)]
+
+        # initiate and run SWA
+        self.model.set_weights(mu_r)
+        self.model.train()
+        base_opt = torch.optim.SGD(self.model.parameters(), lr=self.swa_lr1)
+        opt = SWA(base_opt,  swa_freq=self.swa_freq, swa_lr=self.swa_lr2)
+        for i in range(self.swa_epochs):
+             for batch in T_batched:
+                 opt.zero_grad()
+                 X_batched = [b[0] for b in batch]
+                 p_batched = torch.tensor([b[1] for b in batch])
```

Comment:
Change p to tensor directly, and dont import numpy.

Reply:
p is now a tensor.